scripts/add_headers.sh

```diff
@@ -0,0 +1,36 @@
+#!/usr/bin/env bash
+# Copyright 2023 IBM Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+if ! hash addlicense 2>/dev/null; then
+  go install github.com/google/addlicense@latest
+fi
+
+# adds a new line between the header and the package keyword when it does not exists.
```

Comment:
the `gsed` alias needs one more line to work:

```shell
# set interactive mode to enable defining a gsed alias
shopt -s expand_aliases
```

Reply:
☝🏻 re: [Test (lint) PR check failure](https://github.com/kserve/modelmesh-runtime-adapter/actions/runs/8101930321/job/22539037061#step:5:12)

```
Run ./scripts/develop.sh make fmt
./scripts/add_headers.sh
go: downloading github.com/google/addlicense v1.1.1
go: downloading github.com/bmatcuk/doublestar/v4 v4.0.2
go: downloading golang.org/x/sync v0.0.0-20190911185100-cd5d95a43a6e
find: 'gsed': No such file or directory
find: 'gsed': No such file or directory
find: 'gsed': No such file or directory
...
find: 'gsed': No such file or directory
find: 'gsed': No such file or directory
find: 'gsed': No such file or directory
./scripts/add_headers.sh: line 35: /bin/addlicense: No such file or directory
make: *** [Makefile:64: add_license_headers] Error 127
Error: Process completed with exit code 2.
```

Reply:
Just giving that a spin, turns out `find ... -exec ` cannot find the exported `gsed` alias, so my earlier comment wont make it work.

Here are 3 or 4 alternative that do work for me on Mac and Linux. You pick? :-)

### Option 1: slow `for file in $(find ...)`

```Bash
# set interactive mode to enable defining a gsed alias
shopt -s expand_aliases

# we use sed to make in-file text replacements, but sed works differently depending on the version
if ! sed -i '1s/^/test/' $(mktemp) 2> /dev/null;
then
  alias gsed="sed -i ''"   # macOS (BSD) version of sed
else
  alias gsed="sed -i"      # POSIX compliant version of sed
fi
export gsed

# adds a new line between the header and the package keyword when it does not exists.
for file in $(find . -type f -name "*.go"); do
  echo $file
  gsed \
    -e '/\/\/ limitations under the License\.$/{' \
    -e 'N' \
    -e 's/\(\/\/ limitations under the License.\)\n\(package\)/\1\n\n\2/;}' \
    $file
done
```

### Option 2: export a bash function that can use the `gsed` alias

```Bash
# set interactive mode to enable defining a gsed alias
shopt -s expand_aliases

# we use sed to make in-file text replacements, but sed works differently depending on the version
if ! sed -i '1s/^/test/' $(mktemp) 2> /dev/null; then
    # macOS (BSD) version of sed
    alias gsed="sed -i ''"
else
    # POSIX compliant version of sed
    alias gsed="sed -i"
fi
export gsed

# adds a new line between the header and the package keyword when it does not exists.
add_empty_lines () {
  echo "$1"
  gsed \
    -e '/\/\/ limitations under the License\.$/{' \
    -e 'N' \
    -e 's/\(\/\/ limitations under the License.\)\n\(package\)/\1\n\n\2/;}' \
    "$1"
}
export -f add_empty_lines

# make sure license headers are not treated as a go package comment by `go fmt`
find . -type f -name "*.go" -exec bash -c 'add_empty_lines "$0"' {} \;
```


### Option 3: Your original script, with a `if "macOS" then ...`

No need to define `gsed` alias in this one:

```Bash
# adds a new line between the header and the package keyword when it does not exists.
if ! sed -i '1s/^/test/' $(mktemp) 2> /dev/null; then
  # macOS version of sed: sed -i '' ...
  find . -type f -name "*.go" -exec /usr/bin/sed -i '' -e '/\/\/ limitations under the License\.$/{' -e 'N' -e 's/\(\/\/ limitations under the License.\)\n\(package\)/\1\n\n\2/;}' {} \;
else
  # Posix version of sed: sed -i ...
  find . -type f -name "*.go" -exec /usr/bin/sed -i -e '/\/\/ limitations under the License\.$/{' -e 'N' -e 's/\(\/\/ limitations under the License.\)\n\(package\)/\1\n\n\2/;}' {} \;
fi

# add license to new files that does not contains it
shopt -s globstar
addlicense  -v -c "IBM Corporation" -l=apache Dockerfile **/*.go scripts
```

Reply:
Oh, I didn't see the CI failure =/

Reply:
Guess the first option is fine :)
thanks for look into it.